src/libmongoc/src/mongoc/mongoc-openssl.c

```diff
@@ -477,6 +513,142 @@ _get_issuer (X509 *cert, STACK_OF (X509) * chain)
    RETURN (issuer);
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+/* OpenSSL 1.1.0+ has conveniences that we polyfill in older OpenSSL versions.
+ */
+
+STACK_OF (X509) * _get_verified_chain (SSL *ssl)
+{
+   return SSL_get0_verified_chain (ssl);
+}
+
+void _free_verified_chain (STACK_OF (X509) * verified_chain)
+{
+   /* _get_verified_chain does not return a copy. Do nothing. */
+   return;
+}
+
+const STACK_OF (X509_EXTENSION) * _get_extensions (const X509 *cert)
+{
+   return X509_get0_extensions (cert);
+}
+
+#else
+/* Polyfill functionality for pre 1.1.0 OpenSSL */
+
+STACK_OF (X509) * _get_verified_chain (SSL *ssl)
+{
+   X509_STORE *store = NULL;
+   X509 *peer = NULL;
+   STACK_OF (X509) *peer_chain = NULL;
+   X509_STORE_CTX *store_ctx = NULL;
+   STACK_OF (X509) *verified_chain = NULL;
+
+   /* Get the certificate the server presented. */
+   peer = SSL_get_peer_certificate (ssl);
+   /* Get the chain of certificates the server presented. This is not a verified
+    * chain. */
+   peer_chain = SSL_get_peer_cert_chain (ssl);
+   store = SSL_CTX_get_cert_store (SSL_get_SSL_CTX (ssl));
+   store_ctx = X509_STORE_CTX_new ();
+   if (!X509_STORE_CTX_init (store_ctx, store, peer, peer_chain)) {
+      MONGOC_ERROR ("failed to initialize X509 store");
+      goto fail;
+   }
+
+   if (X509_verify_cert (store_ctx) <= 0) {
+      MONGOC_ERROR ("failed to obtain verified chain");
+      goto fail;
+   }
+
+   verified_chain = X509_STORE_CTX_get1_chain (store_ctx);
+
+fail:
+   X509_free (peer);
+   X509_STORE_CTX_free (store_ctx);
+   return verified_chain;
+}
+
+/* On OpenSSL < 1.1.0, this chain isn't attached to the SSL session, so we need
+ * it to dispose of itself. */
+void _free_verified_chain (STACK_OF (X509) * verified_chain)
+{
+   if (!verified_chain) {
+      return;
+   }
+   sk_X509_pop_free (verified_chain, X509_free);
+}
+
+const STACK_OF (X509_EXTENSION) * _get_extensions (const X509 *cert)
+{
+   return cert->cert_info->extensions;
+}
+#endif /* OPENSSL_VERSION_NUMBER */
+
+
+#define TLSFEATURE_STATUS_REQUEST 5
+
+/* Parse just enough of a DER encoded data to check if a SEQUENCE of INTEGER
+ * contains the status_request extension (5). There are only five tlsfeature
+ * extension types, so this only handles the case that the sequence's length is
+ * representable in one byte, and that each integer is representable in one
+ * byte. */
+static bool
+_sequence_has_status_request (const uint8_t *data, int length)
+{
+   int i;
+
+   /* Expect a sequence type, with a sequence length representable in one byte.
+    */
+   if (length < 3 || data[0] != 0x30 || data[1] >= 127) {
+      MONGOC_ERROR ("malformed tlsfeature extension sequence");
+      return false;
+   }
+
+   for (i = 2; i < length; i += 3) {
+      /* Expect an integer, representable in one byte. */
+      if (length < i + 3 || data[i] != 0x02 || data[i + 1] >= 127) {
```

Comment:
You need to validate the length of the INTEGER is 1 byte otherwise your i+3 logic will start reading the wrong data.

Imagine if the first INTEGER was 2 bytes instead of 1. Your iterator logic would be completely broken.

The JS viewer is pretty cool. Here is my ASN.1 reference - https://www.itu.int/rec/T-REC-X.690-201508-I

Reply:
Ah, good catch! I was checking for a multi-byte length, but not a multi-byte integer. And thanks for the reference.

This was also ripe for unit testing, so I added tests.